crypto/comp/c_zlib.c

```diff
@@ -238,27 +235,34 @@ COMP_METHOD *COMP_zlib(void)
                 = (deflateInit__ft) DSO_bind_func(zlib_dso, "deflateInit_");
             p_zError = (zError__ft) DSO_bind_func(zlib_dso, "zError");
 
-            if (p_compress && p_inflateEnd && p_inflate
-                && p_inflateInit_ && p_deflateEnd
-                && p_deflate && p_deflateInit_ && p_zError)
-                zlib_loaded++;
-
-            if (!OPENSSL_init_crypto(OPENSSL_INIT_ZLIB, NULL)) {
+            if (p_compress == NULL || p_inflateEnd == NULL
+                    || p_inflate == NULL || p_inflateInit_ == NULL
+                    || p_deflateEnd == NULL || p_deflate == NULL
+                    || p_deflateInit_ == NULL || p_zError == NULL
+                    || !OPENSSL_init_crypto(OPENSSL_INIT_ZLIB, NULL)) {
```

Comment:
I hesitate at this. Calling OPENSSL_init_crypto (which calls various RUN_ONCE functions) from inside a RUN_ONCE function seems like asking for deadlock (although I'm fairly sure we do this in other places so possibly not). I actually wonder whether the OSSL_INIT_ZLIB option is even worth having, since all it does it register ossl_comp_zlib_cleanup() to be called at close down. We could just call it unconditionally instead.

Reply:
No deadlocks during test runs.

Reply:
Right...but I'm thinking more about whether this is a pattern we want to see, or whether we should avoid it "in principle". It seems it would not be too difficult to avoid here.

Reply:
All `OpenSSL_init_crypto()` does is set a flag. The ZLIB cleanup code is pretty trivial.

Reply:
It also runs a number of other RUN_ONCE functions (ossl_init_base, ossl_init_register_atexit, ossl_init_load_crypto_nodelete)

Reply:
I'm in the middle of a fixup commit.

Reply:
> It also runs a number of other RUN_ONCE functions (ossl_init_base, ossl_init_register_atexit, ossl_init_load_crypto_nodelete)

I meant for ZLIB.

Reply:
> I meant for ZLIB.

So did I. Calling `OSSL_init_crypto` with `OPENSSL_INIT_ZLIB`, will also run all the above mentioned RUN_ONCE functions as well as the zlib one.

Reply:
As far as ZLIB is concerned, the only thing it tries to do is set a flag so that cleanup is performed; if cleanup is made unconditional, then no need to call this. I'm in violent agreement with you...